rollup/pipeline/pipeline.go

```diff
@@ -0,0 +1,412 @@
+				return
+			}
+
+			if tx.IsL1MessageTx() && tx.AsL1MessageTx().QueueIndex != p.nextL1MsgIndex {
+				// Continue, we might still be able to include some L2 messages
+				resCh <- ErrUnexpectedL1MessageIndex
+				continue
+			}
+
+			if !tx.IsL1MessageTx() && !p.chain.Config().Scroll.IsValidBlockSize(p.blockSize+tx.Size()) {
+				// can't fit this txn in this block, silently ignore and continue looking for more txns
+				resCh <- nil
+				continue
+			}
+
+			// Start executing the transaction
+			p.state.SetTxContext(tx.Hash(), p.txs.Len())
+			receipt, trace, err := p.traceAndApply(tx)
+
+			if p.txs.Len() == 0 && tx.IsL1MessageTx() && err != nil {
+				// L1 message errored as the first txn, skip
+				p.nextL1MsgIndex = tx.AsL1MessageTx().QueueIndex + 1
+			}
+
+			if err == nil {
+				// Everything ok, collect the logs and shift in the next transaction from the same account
+				p.coalescedLogs = append(p.coalescedLogs, receipt.Logs...)
+				p.txs = append(p.txs, tx)
+				p.receipts = append(p.receipts, receipt)
+
+				if !tx.IsL1MessageTx() {
+					// only consider block size limit for L2 transactions
+					p.blockSize += tx.Size()
+				} else {
+					p.nextL1MsgIndex = tx.AsL1MessageTx().QueueIndex + 1
+				}
+
+				stallStart := time.Now()
+				select {
+				case incrementCh <- &PendingBlockIncrement{
+					LastTrace:      trace,
+					NextL1MsgIndex: p.nextL1MsgIndex,
+
+					Header:        types.CopyHeader(&p.Header),
+					State:         p.state.Copy(),
+					Txs:           p.txs,
+					Receipts:      p.receipts,
+					CoalescedLogs: p.coalescedLogs,
+				}:
+				case tx = <-txsIn:
+					if tx != nil {
+						panic("shouldn't have happened")
+					}
+					// next stage terminated and caller terminated us as well
+					return
+				}
+				applyStallTimer.UpdateSince(stallStart)
+			}
+			if err != nil && trace != nil {
+				err = &ErrorWithTrace{
+					Trace: trace,
+					err:   err,
+				}
+			}
+			applyTimer.UpdateSince(applyStart)
+			resCh <- err
+		}
+	}()
+	return resCh, incrementCh, nil
+}
+
+type Result struct {
+	OverflowingTx    *types.Transaction
+	OverflowingTrace *types.BlockTrace
+	CCCErr           error
+
+	Rows       *types.RowConsumption
+	FinalBlock *PendingBlockIncrement
+}
+
+func (p *Pipeline) cccStage(increments <-chan *PendingBlockIncrement, deadline time.Time) chan *Result {
+	p.ccc.Reset()
+	resultCh := make(chan *Result)
+	var lastIncrement *PendingBlockIncrement
+	var lastAccRows *types.RowConsumption
+	var deadlineReached bool
+
+	go func() {
+		defer func() {
+			close(resultCh)
+			lifetimeTimer.UpdateSince(p.start)
+		}()
+		for {
+			idleStart := time.Now()
+			select {
+			case <-time.After(time.Until(deadline)):
+				cccIdleTimer.UpdateSince(idleStart)
+				if lastIncrement != nil {
+					resultCh <- &Result{
+						Rows:       lastAccRows,
+						FinalBlock: lastIncrement,
+					}
+					return
+				}
+				deadlineReached = true
+				deadline = time.Now().Add(time.Hour)
```

Comment:
Why do we add 1 hour here?

Reply:
It is hacky way of making sure that deadline case is not triggered repeatedly after the deadline is hit.

Reply:
Makes sense. Maybe it's better to add a comment about this.

Reply:
done https://github.com/scroll-tech/go-ethereum/pull/735/commits/9d1147ad29fff58c90b7ad63fe0bad4822aca24a